File: v1/src/main/java/com/api/v1/customers/utils/CustomerFinderImpl.java
package com.api.v1.customers.utils;

import com.api.v1.common.*;
import com.api.v1.customers.domain.Customer;
import com.api.v1.customers.domain.CustomerCrudRepository;
import com.api.v1.customers.exceptions.CustomerNotFoundException;
import org.springframework.stereotype.Component;

@Component
public final class CustomerFinderImpl implements CustomerFinder {

    private final CustomerCrudRepository repository;

    public CustomerFinderImpl(CustomerCrudRepository repository) {
        this.repository = repository;
    }

    public Customer findById(@ObjectId String id) {
        return repository
                .findById(id)
                .orElseThrow(() -> new CustomerNotFoundException(id));
    }
}

File: v2/src/main/kotlin/com/api/v2/customers/utils/CustomerFinderImpl.kt
package com.api.v2.customers.utils

import com.api.v2.customers.Customer
import com.api.v2.customers.domain.CustomerCrudRepository
import com.api.v2.customers.exceptions.CustomerNotFoundException
import com.api.v2.customers.CustomerFinder
import org.springframework.stereotype.Component

@Component
class CustomerFinderImpl(val repository: CustomerCrudRepository): CustomerFinder {

    override fun findById(id: String): Customer {
        val foundCustomer = repository.findById(id)
        if (foundCustomer.isEmpty) {
            throw CustomerNotFoundException(id)
        }
        return foundCustomer.get()
    }
}